File: scripts/build-mlb-tools.py
"""
build-mlb-tools.py
Computes MLB career tool grades — training targets for the prospect regression.
Each season z-scored against MLB|YYYY norms, weighted by PA/IP, rolled to career.
Output: data/model/mlb-tools.json
"""
import json, os, glob, math
from collections import defaultdict

BASE       = os.environ.get('DATA_BASE', os.path.expanduser('~/Desktop/fantasy-baseball/data'))
PLAYERS    = os.path.join(BASE, 'players.json')
HIST_DIR   = os.path.join(BASE, 'history')
NORMS_PATH = os.path.join(BASE, 'model', 'norms.json')
OUTPUT     = os.path.join(BASE, 'model', 'mlb-tools.json')

MIN_SEASONS = 1
MIN_PA      = 100
MIN_IP      = 50.0
MIN_GS      = 5

with open(PLAYERS) as f: players = json.load(f)
with open(NORMS_PATH) as f: norms = json.load(f)

history = defaultdict(list)
for path in sorted(glob.glob(os.path.join(HIST_DIR, '*.json'))):
    fname = os.path.basename(path).replace('.json','')
    if not fname.isdigit(): continue
    year = int(fname)
    with open(path) as f:
        for pid, seasons in json.load(f).items():
            for s in seasons:
                history[pid].append({**s, 'year': year})

VALID_YEARS = {2015,2016,2017,2018,2019,2021,2022,2023,2024,2025,2026}

def ip_to_float(ip):
    try:
        parts = str(ip).split('.')
        full = int(parts[0])
        thirds = int(parts[1]) if len(parts) > 1 else 0
        return full + thirds / 3
    except: return 0.0

def zscore(val, norm_entry, stat, invert=False):
    if norm_entry is None: return None
    s = norm_entry.get(stat)
    if not s or s['stdev'] == 0: return None
    z = (val - s['mean']) / s['stdev']
    return -z if invert else z

def career_hitter(pid):
    mlb = [s for s in history.get(str(pid), [])
           if s.get('level') == 'MLB' and s.get('type') == 'hitting'
           and s['year'] in VALID_YEARS and (s.get('pa') or 0) >= MIN_PA]
    if len(mlb) < MIN_SEASONS: return None

    stat_sums = defaultdict(float)
    total_pa  = 0.0

    for s in mlb:
        pa   = s['pa']
        ab   = s.get('ab') or 0
        h    = s.get('h') or 0
        bb   = s.get('bb') or 0
        so   = s.get('so') or 0
        slg  = float(s.get('slg') or 0)
        hbp  = s.get('hbp') or 0
        sb   = s.get('sb') or 0

        avg  = h / ab if ab > 0 else 0
        iso  = slg - avg
        k_pct  = so / pa
        bb_pct = bb / pa
        tob    = h + bb + hbp
        sb_rate = sb / tob if tob > 0 else 0

        n = norms.get(f"MLB|{s['year']}", {}).get('hitters')

        z_avg    = zscore(avg,     n, 'avg')
        z_k      = zscore(k_pct,  n, 'k_pct',  invert=True)
        z_bb     = zscore(bb_pct, n, 'bb_pct')
        z_iso    = zscore(iso,    n, 'iso')
        z_sb     = zscore(sb_rate,n, 'sb_rate')

        for key, z in [('avg',z_avg),('k_pct',z_k),('bb_pct',z_bb),('iso',z_iso),('sb_rate',z_sb)]:
            if z is not None:
                stat_sums[key] += z * pa

        total_pa += pa

    if total_pa == 0: return None

    career_z = {k: v / total_pa for k, v in stat_sums.items()}

    def to_plus(z): return round(100 + z * 15)

    hit   = round((to_plus(career_z.get('avg',0)) +
                   to_plus(career_z.get('k_pct',0)) +
                   to_plus(career_z.get('bb_pct',0))) / 3)
    power = to_plus(career_z.get('iso', 0))
    speed = to_plus(career_z.get('sb_rate', 0))

    return {
        'type': 'hitter',
        'hit': hit, 'power': power, 'speed': speed,
        '_seasons': len(mlb), '_pa': round(total_pa),
        '_z': {k: round(v,3) for k,v in career_z.items()},
    }

def career_pitcher(pid):
    mlb = [s for s in history.get(str(pid), [])
           if s.get('level') == 'MLB' and s.get('type') == 'pitching'
           and s['year'] in VALID_YEARS
           and ip_to_float(s.get('ip',0)) >= MIN_IP
           and (s.get('gs') or 0) >= MIN_GS]
    if len(mlb) < MIN_SEASONS: return None

    stat_sums = defaultdict(float)
    total_ip  = 0.0

    for s in mlb:
        ip  = ip_to_float(s.get('ip', 0))
        bf  = s.get('bf') or 0
        so  = s.get('so') or 0
        bb  = s.get('bb') or 0

        if bf == 0: continue

        k_pct  = so / bf
        bb_pct = bb / bf

        n = norms.get(f"MLB|{s['year']}", {}).get('pitchers')

        z_k  = zscore(k_pct,  n, 'k_pct')
        z_bb = zscore(bb_pct, n, 'bb_pct', invert=True)

        for key, z in [('k_pct', z_k), ('bb_pct', z_bb)]:
            if z is not None:
                stat_sums[key] += z * ip

        total_ip += ip

    if total_ip == 0: return None

    career_z = {k: v / total_ip for k, v in stat_sums.items()}

    def to_plus(z): return round(100 + z * 15)

    stuff   = to_plus(career_z.get('k_pct', 0))
    control = to_plus(career_z.get('bb_pct', 0))

    return {
        'type': 'pitcher',
        'stuff': stuff, 'control': control,
        '_seasons': len(mlb), '_ip': round(total_ip),
        '_z': {k: round(v,3) for k,v in career_z.items()},
    }

output = {}
skipped = 0

for pid, p in players.items():
    mlbam = p.get('mlbam_id')
    if not mlbam: continue
    is_p = 'P' in (p.get('positions') or '')
    result = career_pitcher(mlbam) if is_p else career_hitter(mlbam)
    if not result:
        skipped += 1
        continue
    result['name'] = p['name']
    output[str(mlbam)] = result

print(f'Qualifying hitters:  {sum(1 for v in output.values() if v["type"]=="hitter")}')
print(f'Qualifying pitchers: {sum(1 for v in output.values() if v["type"]=="pitcher")}')
print(f'Skipped (insufficient): {skipped}')

os.makedirs(os.path.dirname(OUTPUT), exist_ok=True)
with open(OUTPUT, 'w') as f: json.dump(output, f, indent=2)
print(f'Wrote {len(output)} tool grades → model/mlb-tools.json')

File: scripts/build-mlb-tools.js
// build-mlb-tools.js
// Node replacement for build-mlb-tools.py.
// Computes MLB career tool grades (training targets for the prospect regression)
// using the canonical formula in lib/score-tools.js — same logic the arc and
// build-scores.js use, so all surfaces stay in lockstep.
//
// Output: data/model/mlb-tools.json  (same shape as the Python script's output)

const fs   = require('fs');
const path = require('path');
const os   = require('os');
const { scoreMlbTool } = require('../lib/score-tools');

const BASE         = process.env.DATA_BASE || path.join(os.homedir(), 'Desktop/fantasy-baseball/data');
const PLAYERS_PATH = path.join(BASE, 'players.json');
const HIST_DIR     = path.join(BASE, 'history');
const NORMS_PATH   = path.join(BASE, 'model/norms.json');
const OUTPUT       = path.join(BASE, 'model/mlb-tools.json');

const VALID_YEARS  = new Set([2015,2016,2017,2018,2019,2021,2022,2023,2024,2025,2026]);
const CURRENT_YEAR = new Date().getFullYear();
const MIN_PA       = 100;
const MIN_IP       = 50.0;

const players = JSON.parse(fs.readFileSync(PLAYERS_PATH, 'utf8'));
const norms   = JSON.parse(fs.readFileSync(NORMS_PATH, 'utf8'));

const history = {};
for (const file of fs.readdirSync(HIST_DIR).sort()) {
  const m = file.match(/^(\d{4})\.json$/);
  if (!m) continue;
  const year = parseInt(m[1]);
  const data = JSON.parse(fs.readFileSync(path.join(HIST_DIR, file), 'utf8'));
  for (const [pid, seasons] of Object.entries(data)) {
    if (!history[pid]) history[pid] = [];
    for (const s of seasons) history[pid].push({ ...s, year });
  }
}

function ipToFloat(ip) {
  const p = String(ip || 0).split('.');
  return parseInt(p[0] || '0') + parseInt(p[1] || '0') / 3;
}

function bucketFromSeason(s) {
  const ipP = String(s.ip || '0').split('.');
  return {
    year: s.year, level: 'MLB',
    ab:  s.ab  || 0, bb: s.bb || 0, hbp: s.hbp || 0,
    so:  s.so  || 0, h:  s.h  || 0,
    xbh: (s.doubles || 0) + (s.triples || 0) + (s.hr || 0),
    sb:  s.sb  || 0,
    tb:  s.tb  || Math.round((parseFloat(s.slg || '0') || 0) * (s.ab || 0)),
    bf:  s.bf  || 0,
    ipOuts: (parseInt(ipP[0] || '0') * 3) + parseInt(ipP[1] || '0'),
  };
}

function careerHitter(pid) {
  const mlb = (history[String(pid)] || []).filter(s =>
    s.level === 'MLB' && s.type === 'hitting' && VALID_YEARS.has(s.year) && (s.pa || 0) >= MIN_PA
  );
  if (!mlb.length) return null;
  const buckets = mlb.map(bucketFromSeason);
  const opts = { norms, isPit: false, referenceYear: CURRENT_YEAR };
  const totalPA = mlb.reduce((sum, s) => sum + (s.pa || 0), 0);
  return {
    type: 'hitter',
    hit:   scoreMlbTool(buckets, 'hit',   opts).score,
    power: scoreMlbTool(buckets, 'power', opts).score,
    speed: scoreMlbTool(buckets, 'speed', opts).score,
    _seasons: mlb.length,
    _pa: Math.round(totalPA),
  };
}

function careerPitcher(pid) {
  const mlb = (history[String(pid)] || []).filter(s =>
    s.level === 'MLB' && s.type === 'pitching' && VALID_YEARS.has(s.year)
  );
  if (!mlb.length) return null;
  const totalIP = mlb.reduce((sum, s) => sum + ipToFloat(s.ip), 0);
  if (totalIP < MIN_IP) return null;
  const buckets = mlb.map(bucketFromSeason).filter(b => b.bf > 0);
  if (!buckets.length) return null;
  const opts = { norms, isPit: true, referenceYear: CURRENT_YEAR };
  return {
    type: 'pitcher',
    stuff:   scoreMlbTool(buckets, 'stuff',   opts).score,
    control: scoreMlbTool(buckets, 'control', opts).score,
    _seasons: mlb.length,
    _ip: Math.round(totalIP),
  };
}

function isTwoWayPositions(positions) {
  const pos = (positions || '').split(',').map(s => s.trim());
  const hasArm = pos.some(p => p === 'SP' || p === 'RP' || p === 'P');
  const hasBat = pos.some(p => p !== 'SP' && p !== 'RP' && p !== 'P');
  return { hasArm, hasBat };
}

const output = {};
let skipped = 0;

for (const [, p] of Object.entries(players)) {
  const mlbam = p.mlbam_id;
  if (!mlbam) continue;
  const { hasArm, hasBat } = isTwoWayPositions(p.positions);

  let result;
  if (hasArm && hasBat) {
    const h   = careerHitter(mlbam);
    const pit = careerPitcher(mlbam);
    if (!h && !pit) { skipped++; continue; }
    result = { type: 'two-way', name: p.name };
    if (h)   { result.hit = h.hit; result.power = h.power; result.speed = h.speed; result._pa = h._pa; }
    if (pit) { result.stuff = pit.stuff; result.control = pit.control; result._ip = pit._ip; }
  } else if (hasArm) {
    result = careerPitcher(mlbam);
    if (!result) { skipped++; continue; }
    result.name = p.name;
  } else {
    result = careerHitter(mlbam);
    if (!result) { skipped++; continue; }
    result.name = p.name;
  }

  output[String(mlbam)] = result;
}

const counts = { hitter: 0, pitcher: 0, 'two-way': 0 };
for (const v of Object.values(output)) counts[v.type] = (counts[v.type] || 0) + 1;
console.log(`Qualifying hitters:  ${counts.hitter}`);
console.log(`Qualifying pitchers: ${counts.pitcher}`);
console.log(`Qualifying two-way:  ${counts['two-way']}`);
console.log(`Skipped (insufficient): ${skipped}`);

fs.mkdirSync(path.dirname(OUTPUT), { recursive: true });
fs.writeFileSync(OUTPUT, JSON.stringify(output, null, 2));
console.log(`Wrote ${Object.keys(output).length} tool grades → model/mlb-tools.json`);
